component/stream/stream.go

```diff
@@ -245,57 +176,74 @@ func (st *Stream) OpenFile(options internal.OpenFileOptions) (*handlemap.Handle,
 		handle = handlemap.NewHandle(options.Name)
 	}
 	if !st.streamOnly {
-		fileKey := options.Name
-		if st.cacheOnFileHandle {
-			handle.ID = handlemap.HandleID(nextHandleID.Inc())
-			fileKey = strconv.FormatUint((uint64(handle.ID)), 10)
+		if st.cache.openHandles >= st.cache.handleLimit {
+			err = errors.New("handle limit exceeded")
+			log.Err("Stream::OpenFile : error %s [%s]", options.Name, err.Error())
+			return handle, err
 		}
-
-		flock := st.fileKeyLocks.Get(fileKey)
-		flock.Lock()
-		st.streamCache.addFileKey(fileKey)
-		block, exists, _ := st.fetchBlock(fileKey, handle, 0)
+		atomic.AddInt32(&st.cache.openHandles, 1)
+		cacheObj := handlemap.Cache{
+			DataBuffer: common.NewLRUCache(int64(st.cache.bufferSizePerHandle)),
+		}
+		handle.CacheObj = &cacheObj
+		block, exists, _ := st.getBlock(handle, 0)
 		// if it exists then we can just RUnlock since we didn't manipulate its data buffer
 		st.unlockBlock(block, exists)
-		st.streamCache.incrementHandles(fileKey)
-		flock.Unlock()
 	}
 	return handle, err
 }
 
-func (st *Stream) fetchBlock(fileKey string, handle *handlemap.Handle, offset int64) (*cacheBlock, bool, error) {
-	block, exists := st.streamCache.getBlock(fileKey, offset, handle.Size)
-	if !exists {
+func (st *Stream) getBlock(handle *handlemap.Handle, offset int64) (*common.CacheBlock, bool, error) {
+	blockSize := st.cache.blockSize
+	blockKeyObj := offset
+	handle.CacheObj.Lock()
+	block, found := handle.CacheObj.DataBuffer.Get(blockKeyObj)
+	if !found {
+		if (offset + blockSize) > handle.Size {
```

Comment:
Can we remove all this block allocations related stuff from here. May be just like handle and policy we can have a cache object file and allocation of block and all can be done inside. That simplifies the outside logic where you just call Get()/Put() on handle and if required it can allocate the required block inside and return back a locked block object.

Reply:
This will also help us keep a finer control over locks so that in no code change we accidently mess it up here.

Reply:
Even the unlock of block can be moved inside the cache layer instead of having a wrapper function here.

Reply:
I think youre describing the read.go and read_write.go, where stream -> read.go/read_write.go -> lru_policy.go. Where implementation of block etc is gonna be specific to read only or read write mode. This shall be in the pr for write which im working on. In the meantime this is sufficient changes in one PR. For the locking of block ill move it outside of lru_policy